lib/orderbook/OrderBook.ts

```diff
@@ -479,51 +479,61 @@ class OrderBook extends EventEmitter {
-      const availableQuantity = order.quantity - order.hold;
-      if (availableQuantity) {
-        this.removeOwnOrder(orderIdentifier.id, orderIdentifier.pairId, availableQuantity);
+
+    let remainingQuantityToRemove = quantityToRemove || order.quantity;
+
+    const failedHandler = (deal: SwapDeal) => {
+      if (deal.orderId === orderIdentifier.id) {
+        // remove the portion that failed now that it's not on hold
+        const quantityToRemove = Math.min(deal.quantity!, remainingQuantityToRemove);
+        this.removeOwnOrder(orderIdentifier.id, orderIdentifier.pairId, quantityToRemove);
+        cleanup(quantityToRemove);
       }
+    };
 
-      const cleanup = (quantity: number) => {
-        remainingHold -= quantity;
-        this.logger.debug(`removed hold of ${quantity} on local order ${localId}, ${remainingHold} remaining`);
-        if (remainingHold === 0) {
-          // we can stop listening for swaps once all holds are cleared
-          this.swaps!.removeListener('swap.failed', failedHandler);
-          this.swaps!.removeListener('swap.paid', paidHandler);
-        }
-      };
+    const paidHandler = (result: SwapSuccess) => {
+      if (result.orderId === orderIdentifier.id) {
+        const quantityToRemove = Math.min(result.quantity, remainingQuantityToRemove);
+        cleanup(quantityToRemove);
+      }
+    };
 
-      const failedHandler = (deal: SwapDeal) => {
-        if (deal.orderId === orderIdentifier.id) {
-          // remove the portion that failed now that it's not on hold
-          this.removeOwnOrder(orderIdentifier.id, orderIdentifier.pairId, deal.quantity!);
-          cleanup(deal.quantity!);
-        }
-      };
+    const cleanup = (quantity: number) => {
+      remainingQuantityToRemove -= quantity;
+      this.logger.debug(`removed hold of ${quantity} on local order ${localId}, ${remainingQuantityToRemove} remaining`);
+      if (remainingQuantityToRemove === 0) {
+        // we can stop listening for swaps once all holds are cleared
+        this.swaps!.removeListener('swap.failed', failedHandler);
+        this.swaps!.removeListener('swap.paid', paidHandler);
+      }
+    };
 
-      const paidHandler = (result: SwapSuccess) => {
-        if (result.orderId === orderIdentifier.id) {
-          cleanup(result.quantity);
-        }
-      };
+    if (remainingQuantityToRemove > order.quantity) {
+      // quantity to be removed can't be higher than order's quantity.
+      throw errors.QUANTITY_DOES_NOT_MATCH(remainingQuantityToRemove, order.quantity);
+    }
+
+    const removableQuantity = order.quantity - order.hold;
+    if (remainingQuantityToRemove <= removableQuantity) {
+      this.removeOwnOrder(orderIdentifier.id, orderIdentifier.pairId, quantityToRemove);
+      remainingQuantityToRemove = 0;
+    } else {
+      // we can't remove the entire amount because of a hold on the order
```

Comment:
perhaps put `failedHandler` / `paidHandler` / `cleanup` inside of this clause, since they are only relevant on this case.